### ISSUE TITLE
False positive: no-member on "property"-like decorator

### ISSUE THREAD

**Issue by TBoshoven** [pylint-dev/pylint#1127]
### Steps to reproduce
1. Have Pylint (1.6.4) evaluate the following snippet:
   
   ``` python
   class prop(object):
       def __init__(self, fn):
           self.value = 5
   
   class my_class(object):
       @prop
       def my_property(self):
           pass
   
   print(my_class().my_property.value)
   ```
### Current behavior

Pylint returns the following, even with `prop` added to `property_classes`:
    `E: 10, 6: Method 'my_property' has no 'value' member (no-member)`
### Expected behavior

No `no-member` errors. A decorated function should not be assumed to be a function.
### `pylint --version` output

```
    No config file found, using default configuration
    pylint 1.6.4, 
    astroid 1.4.8
    Python 3.5.2 (default, Jun 28 2016, 08:46:01) 
    [GCC 6.1.1 20160602]
```

**Comment by ngeraedts** [comment]
I've run into a similar issue with value injection on our unit tests. The following snippet runs without issue under pytest, but throws a `no-member` error under pylint.

```python
from unittest import TestCase

def with_value(Cls):
    Cls_setUp = Cls.setUp

    def _setUp(self):
        self.value = 5
        self.addCleanup(delattr, self, 'value')
        Cls_setUp(self)

    Cls.setUp = _setUp

    return Cls

@with_value
class TestValue(TestCase):
    def test_value(self):
        value = self.value
        self.assertEqual(value, 5)
```

### pylint --version
```
pylint 1.6.4,
astroid 1.4.8
Python 2.7.9 (default, Sep 26 2016, 12:51:32)
[GCC 4.4.6 20120305 (Red Hat 4.4.6-4)]
```

**Comment by craiga** [comment]
Is this a duplicate of #1061 ?

**Comment by Xunnamius** [comment]
Big still bites in 2019

**Comment by clavedeluna** [comment]
Maintainers, I'm wondering if this could be fixed by implemented the same logic `--contextmanager-decorators` and `config.contextmanager_decorators`. If we see a func decorated with a user-defined `property-decorator`, we would not issue `no-member`.

**Comment by Pierre-Sassoulas** [comment]
I was not aware of this option, good job reading the doc. I also like the idea a lot, fixing the decorator understanding require control flow so it's not going to be fixed by tomorrow. This option would be useful short and mid term.

**Comment by clavedeluna** [comment]
@Pierre-Sassoulas in my attempt to add the following to typecheck.py (for no-member)
```
        (
            "property-decorators",
            {
                "default": [],
                "type": "csv",
                "metavar": "<property decorator names>",
                "help": "List of decorators that produce properties. "
                        "Add to this list to register decorators that "
                        "produce valid properties.",
            },
        ),
```

I noticed that we actually already have config option like this in checker.py
```
        (
            "property-classes",
            {
                "default": ("abc.abstractproperty",),
                "type": "csv",
                "metavar": "<decorator names>",
                "help": "List of decorators that produce properties, such as "
                "abc.abstractproperty. Add to this list to register "
                "other decorators that produce valid properties. "
                "These decorators are taken in consideration only for invalid-name.",
            },
        ),
```

Is it a good idea to repurpose `property-classes` for no-member, too? If so, can you give me an example of a config option that's used across different checkers so I can follow that guide?

**Comment by jacobtylerwalls** [comment]
Duplicate of #9246 -- this goes deeper than properties, any assignment provided by a decorator isn't understood
